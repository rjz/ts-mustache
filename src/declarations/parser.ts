import * as M from 'mustache'
import * as Graph from 'ts-directed-graph'
import { SpanType, ParserNode, ValueNode } from './types'

function joinId(prefix: string, name: string): string {
  return [prefix, name].join('_')
}

/**
 * Technically mustache supports cyclic resolution at runtime. This library
 * isn't subtle enough to deal with it (at least for now)
 */
export class Parser extends Graph.DirectedAcyclicGraph<ParserNode> {
  protected _templates: string[] = []

  get templates(): readonly string[] {
    return Object.freeze([...this._templates])
  }

  protected ensureChild<T extends ParserNode>(
    parentNode: ParserNode,
    node: T,
  ): T {
    const { id } = node
    if (!this.has(id)) {
      this.addNode(node)
    }

    this.addEdge(parentNode.id, id)

    return this.getNode(id) as T
  }

  protected ensureValue(parentNode: ParserNode, name: string): ValueNode {
    return this.ensureChild(parentNode, {
      id: joinId(parentNode.id, name),
      type: 'VALUE',
      propertyKey: name,
    })
  }

  /**
   * @todo handle bracket notation (`foo['bar']`) as well?
   * @todo handle lambdas
   */
  protected mergeValue(parentNode: ParserNode, fullName: string): void {
    if (fullName === '.' && parentNode.type === 'SECTION') {
      this.ensureValue(parentNode, '.')
      return
    }

    const [name, ...parts] = fullName.split('.')

    const node = this.ensureValue(parentNode, name)

    if (parts.length > 0) {
      this.mergeValue(node, parts.join('.'))
    }
  }

  /**
   * @todo handle case of partial merged inside section. The current approach
   *       depends on ambient data feeding the partial (which will work for
   *       many templates data/models) but `{{#users}}{>_profile}}{{/users}}`
   *       may not resolve correctly
   */
  protected mergePartial(parentNode: ParserNode, name: string) {
    this.ensureChild(parentNode, {
      id: name,
      type: 'TEMPLATE',
      propertyKey: name,
    })
  }

  protected mergeSection(
    parentNode: ParserNode,
    fullName: string,
    spans: M.TemplateSpans,
  ): void {
    const [name, ...parts] = fullName.split('.')

    if (parts.length > 0) {
      const node = this.ensureValue(parentNode, name)
      this.mergeSection(node, parts.join('.'), spans)
      return
    }

    const node = this.ensureChild(parentNode, {
      id: joinId(parentNode.id, name),
      type: 'SECTION',
      propertyKey: name,
    })

    this.mergeSpans(node, spans)
  }

  protected mergeSpans(parentNode: ParserNode, spans: M.TemplateSpans) {
    for (const span of spans) {
      const [type, name, ...params] = span

      const t: M.TemplateSpanType = type
      switch (t) {
        case SpanType.ESCAPED_VALUE:
        case SpanType.UNESCAPED_VALUE:
          this.mergeValue(parentNode, name)
          break
        case SpanType.PARTIAL:
          this.mergePartial(parentNode, name)
          break
        case SpanType.SECTION:
        case SpanType.INVERTED:
          if (!Array.isArray(params[2])) {
            // We're really just coercing types. If you're hitting this,
            // Mustache's parsing API has been updated in an unexpected way.
            throw new Error('Missing spans for section')
          }

          this.mergeSection(parentNode, name, params[2])
          break
        case SpanType.COMMENT:
        case SpanType.EQUAL:
        case SpanType.RAW_VALUE:
          break
      }
    }
  }

  addTemplate(templateName: string, spans: M.TemplateSpans) {
    const root = {
      type: 'TEMPLATE',
      id: templateName,
      propertyKey: templateName,
    } as const

    // a `TemplateNode` may already exist from a `PARTIAL` seen previously in
    // another template
    if (!this.has(root.id)) {
      this.addNode(root)
    }

    this.mergeSpans(root, spans)

    this._templates.push(templateName)
  }
}
