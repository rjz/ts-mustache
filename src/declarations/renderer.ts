import { ParserNode, TemplateNode, assertExhaustiveCheck } from './types'
import { Parser } from './parser'

export const utilityTypes = {
  /**
   *  We can't infer type from an untyped template, but we can allow the full
   *  range of valid mustache values
   *
   *  @todo support lambdas (`function` values)
   */
  VALUE: 'type MustacheValue = string | number | boolean',

  /**
   *  Properties in a `RECORD` may be nullable but the record itself must be
   *  passed to the template
   */
  RECORD: 'type MustacheRecord<T> = T',

  /**
   *  A `SECTION` (inverted or otherwise)'s properties are nullable and may or
   *  may not be list types
   */
  SECTION: 'type MustacheSection<T> = T[] | T',
}

type UtilityType = keyof typeof utilityTypes

type ResolutionCandidate =
  | { type: 'VALUE' }
  | { type: 'OPTIONAL' }
  | { type: 'RECORD'; typeName: string }
  | { type: 'SECTION'; typeName: string }

export type Resolution = {
  typeName: string
  candidates: Record<string, ResolutionCandidate[]>
}

export type ResolutionMap = Map<ParserNode['id'], Resolution>

function upperFirst(str: string): string {
  return str[0].toUpperCase() + str.slice(1)
}

function namespaceFor(t: TemplateNode): string {
  return upperFirst(t.propertyKey).replace(/[^a-z0-9_]+/gi, '_')
}

export class Renderer {
  protected resolutions: ResolutionMap = new Map()
  protected utilityTypesUsed = new Set<UtilityType>()

  constructor(protected parsed: Parser) {}

  protected addCandidate(
    resolution: Resolution,
    propertyKey: string,
    candidate: ResolutionCandidate,
  ): void {
    if (!resolution.candidates[propertyKey]) {
      resolution.candidates[propertyKey] = []
    }

    resolution.candidates[propertyKey].push(candidate)
  }

  protected resolveNode(
    id: ParserNode['id'],
    ns: string,
    children: Set<ParserNode>,
  ) {
    let resolution = this.resolutions.get(id)
    if (!resolution) {
      resolution = {
        typeName: ns,
        candidates: {},
      }

      // Make this immediately available to recursive calls
      this.resolutions.set(id, resolution)
    }

    for (const c of children) {
      switch (c.type) {
        case 'SECTION': {
          this.resolveNode(
            c.id,
            [ns, upperFirst(c.propertyKey)].join(''),
            this.parsed.edgesFrom(c.id),
          )

          const section = this.resolutions.get(c.id)!
          if (Object.keys(section.candidates).length === 0) {
            this.addCandidate(resolution, c.propertyKey, { type: 'OPTIONAL' })
            this.resolutions.delete(c.id)
          } else {
            this.addCandidate(resolution, c.propertyKey, {
              type: 'SECTION',
              typeName: section.typeName,
            })
          }

          break
        }
        case 'VALUE': {
          const edges = this.parsed.edgesFrom(c.id)
          if (edges.size === 0) {
            this.addCandidate(resolution, c.propertyKey, { type: 'VALUE' })
          } else {
            this.resolveNode(
              c.id,
              [ns, upperFirst(c.propertyKey)].join(''),
              edges,
            )

            this.addCandidate(resolution, c.propertyKey, {
              type: 'RECORD',
              typeName: this.resolutions.get(c.id)!.typeName,
            })
          }
          break
        }
        case 'TEMPLATE': {
          // We're just going to merge it right into the current node.
          const partialEdges = this.parsed.edgesFrom(c.id)
          this.resolveNode(id, ns, partialEdges)
          break
        }
        default:
          assertExhaustiveCheck(c)
      }
    }

    this.resolutions.set(id, resolution)
  }

  protected resolveTemplate(t: TemplateNode) {
    const ns = namespaceFor(t)
    const children = this.parsed.edgesFrom(t.id)
    this.resolveNode(t.id, ns, children)
  }

  protected propertyString(
    k: string,
    candidates: ResolutionCandidate[],
  ): string {
    let isOptional = false
    const cs: string[] = []
    for (const c of candidates) {
      switch (c.type) {
        case 'RECORD':
          cs.push(`MustacheRecord<${c.typeName}>`)
          this.utilityTypesUsed.add(c.type)
          break
        case 'SECTION':
          isOptional = true
          cs.push(`MustacheSection<${c.typeName}>`) // | MustacheValue
          this.utilityTypesUsed.add(c.type)
          break
        case 'VALUE':
          cs.push(`MustacheValue`)
          this.utilityTypesUsed.add(c.type)
          break
        case 'OPTIONAL':
          isOptional = true
          cs.push(`MustacheValue`)
          this.utilityTypesUsed.add('VALUE')
          break
        default:
          assertExhaustiveCheck(c)
      }
    }

    return `  ${k}${isOptional ? '?' : ''}: ${Array.from(new Set(cs)).join(
      ' & ',
    )}`
  }

  protected resolutionToTypeString(r: Resolution): string {
    const entries = Object.entries(r.candidates)
    const propStr =
      entries.length === 0
        ? ''
        : ['', ...entries.map(([k, p]) => this.propertyString(k, p)), ''].join(
            '\n',
          )

    return `interface ${r.typeName} {${propStr}}`
  }

  toString(): string {
    const templateMap: Record<string, string> = {}

    const templatesVisited = new Set<string>()

    for (const n of this.parsed.nodes()) {
      if (n.type === 'TEMPLATE') {
        this.resolveTemplate(n)
        templateMap[n.id] = this.resolutions.get(n.id)!.typeName
        templatesVisited.add(n.id)
      }
    }

    // Sanity check: ensure that any partials referenced were actually supplied
    // to the parser and rendered
    const parsedTemplates = this.parsed.templates
    for (const t of templatesVisited) {
      if (!parsedTemplates.includes(t)) {
        throw new Error(`Unknown template: ${t}`)
      }
    }

    const output: string[] = []
    for (const r of this.resolutions.values()) {
      output.push(this.resolutionToTypeString(r))
    }

    const utilities: string[] = []
    for (const [k, v] of Object.entries(utilityTypes)) {
      if (this.utilityTypesUsed.has(k as UtilityType)) {
        utilities.push(v)
      }
    }

    output.unshift(...utilities)

    output.push(
      `export type TemplateMap = {\n${Object.entries(templateMap)
        .map(([k, v]) => `  '${k}': ${v},`)
        .join('\n')}\n}`,

      'export type TemplateName = keyof TemplateMap',

      `export const TEMPLATES = [\n${Object.keys(templateMap)
        .map((k) => `  '${k}',`)
        .join('\n')}\n] as const`,
    )

    return output.join('\n\n')
  }
}
