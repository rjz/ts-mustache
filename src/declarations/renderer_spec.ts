import mustache from 'mustache'
import { Parser } from './parser'
import { Resolution, Renderer } from './renderer'

class TestWriter extends Renderer {
  get resolutionMap() {
    return this.resolutions
  }

  findTypeName(typeName: string): Resolution {
    for (const [k, v] of this.resolutions) {
      if (v.typeName === typeName) {
        return v
      }
    }

    throw new Error(`Type "${typeName}" isn't resolved`)
  }

  static resolve(templates: Record<string, string>): TestWriter {
    const parsed = new Parser()

    Object.entries(templates).forEach(([name, template]) => {
      parsed.addTemplate(name, mustache.parse(template))
    })

    const w = new TestWriter(parsed)

    for (const n of parsed.nodes()) {
      if (n.type === 'TEMPLATE') {
        w.resolveTemplate(n)
      }
    }

    return w
  }
}

describe('Parser', function () {
  describe('references', function () {
    it('resolves scalar', function () {
      const template = 'Hello, {{ name }}'
      const w = TestWriter.resolve({ template })
      const templateResolution = w.findTypeName('Template')
      expect(templateResolution.candidates).toEqual({
        name: [{ type: 'VALUE' }],
      })
    })

    it('resolves object reference', function () {
      const template = 'Hello, {{ person.name }}'
      const w = TestWriter.resolve({ template })
      const templateResolution = w.findTypeName('Template')
      expect(templateResolution.candidates).toEqual({
        person: [{ type: 'RECORD', typeName: 'TemplatePerson' }],
      })

      const personResolution = w.findTypeName('TemplatePerson')
      expect(personResolution.candidates).toEqual({
        name: [{ type: 'VALUE' }],
      })
    })
  })

  describe('sections', function () {
    it('resolves object reference', function () {
      const template = '{{#users}}{{name}}{{/users}}'

      const w = TestWriter.resolve({ template })
      const templateResolution = w.findTypeName('Template')

      expect(templateResolution.candidates).toEqual({
        users: [{ type: 'SECTION', typeName: 'TemplateUsers' }],
      })

      const usersResolution = w.findTypeName('TemplateUsers')
      expect(usersResolution.candidates).toEqual({
        name: [{ type: 'VALUE' }],
      })
    })

    it('treats empty sections as SECTIONs', function () {
      const template = '{{#lambda}}nothing to see here but content{{/lambda}}'

      const w = TestWriter.resolve({ template })
      const templateResolution = w.findTypeName('Template')

      expect(templateResolution.candidates).toEqual({
        lambda: [{ type: 'OPTIONAL' }],
      })
    })

    it('resolves conflicting hints', function () {
      const template = `
        {{^users.length}}no users{{/users.length}}
        {{#users}}
          Hello, {{name}}
        {{/users}}
      `

      const w = TestWriter.resolve({ template })
      const templateResolution = w.findTypeName('Template')

      expect(templateResolution.candidates).toEqual({
        users: [{ type: 'RECORD', typeName: 'TemplateUsers' }],
      })

      const usersResolution = w.findTypeName('TemplateUsers')
      expect(usersResolution.candidates).toEqual({
        length: [{ type: 'OPTIONAL' }],
        name: [{ type: 'VALUE' }],
      })
    })
  })

  describe('partials', function () {
    it('resolves partial at root', function () {
      const _profile = '{{person.name}}'
      const template = '{{>_profile}}'

      const w = TestWriter.resolve({ template, _profile })

      const templateResolution = w.findTypeName('Template')
      w.findTypeName('_profile') // no throw == passing test!

      expect(templateResolution.candidates).toEqual({
        person: [{ type: 'RECORD', typeName: 'TemplatePerson' }],
      })
    })

    /**
     * @todo decide how (whether?) to merge nested partials into section and/or
     *       top-level types
     */
    xit('resolves partial inside nested sections', function () {
      const _profile = '{{person.name}}'
      const template =
        '{{#org}}{{#users}}{{>_profile}}{{foo.bar}}{{/users}}{{/org}}'

      const w = TestWriter.resolve({ template, _profile })
      const templateResolution = w.findTypeName('Template')

      // Hoists to top level, too?
      expect(templateResolution.candidates).toEqual({
        org: [{ type: 'SECTION', typeName: 'TemplateOrg' }],
        person: [{ type: 'RECORD', typeName: 'TemplatePerson' }],
      })
    })

    it('throws before generating type with broken reference', function () {
      const template = '{{> missing_partial}}'
      const w = TestWriter.resolve({ template })
      expect(() => w.toString()).toThrow()
    })
  })

  describe('full tests', function () {
    const TEST_PARTIAL = `{{name}} is {{age}} years old`

    const TEST_NAMETAG = `
    {{#person}}{{> _person}}{{/person}} or {{name}} it shouldn't matter
  `

    const TEST_PROFILE = `
  {{> _person}}'s profile
`

    it('works', function () {
      const p = new Parser()
      p.addTemplate('_person', mustache.parse(TEST_PARTIAL))
      p.addTemplate('nametag', mustache.parse(TEST_NAMETAG))
      p.addTemplate('profile', mustache.parse(TEST_PROFILE))

      const t = new Renderer(p)
      const actual = t.toString()

      expect(actual).toMatchSnapshot()
    })

    it('handles nested properties', function () {
      const p = new Parser()
      p.addTemplate('test', mustache.parse('{{ foo.bar }}'))

      const t = new Renderer(p)
      const actual = t.toString()

      expect(actual).toMatchSnapshot()
    })

    it('handles nested properties in sections', function () {
      const p = new Parser()
      p.addTemplate(
        'test',
        mustache.parse('{{# foo.bar }} {{pleaseDontButOk}} {{/ foo.bar}}'),
      )

      const t = new Renderer(p)
      const actual = t.toString()

      expect(actual).toMatchSnapshot()
    })
  })
})
