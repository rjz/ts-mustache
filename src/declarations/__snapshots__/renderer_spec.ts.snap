// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser full tests handles nested properties 1`] = `
"type MustacheValue = string | number | boolean

type MustacheRecord<T> = T

interface Test {
  foo: MustacheRecord<TestFoo>
}

interface TestFoo {
  bar: MustacheValue
}

export type TemplateMap = {
  'test': Test,
}

export type TemplateName = keyof TemplateMap

export const TEMPLATES = [
  'test',
] as const"
`;

exports[`Parser full tests handles nested properties in sections 1`] = `
"type MustacheValue = string | number | boolean

type MustacheRecord<T> = T

type MustacheSection<T> = T[] | T

interface Test {
  foo: MustacheRecord<TestFoo>
}

interface TestFoo {
  bar?: MustacheSection<TestFooBar>
}

interface TestFooBar {
  pleaseDontButOk: MustacheValue
}

export type TemplateMap = {
  'test': Test,
}

export type TemplateName = keyof TemplateMap

export const TEMPLATES = [
  'test',
] as const"
`;

exports[`Parser full tests works 1`] = `
"type MustacheValue = string | number | boolean

type MustacheSection<T> = T[] | T

interface _person {
  name: MustacheValue
  age: MustacheValue
}

interface Nametag {
  person?: MustacheSection<NametagPerson>
  name: MustacheValue
}

interface NametagPerson {
  name: MustacheValue
  age: MustacheValue
}

interface Profile {
  name: MustacheValue
  age: MustacheValue
}

export type TemplateMap = {
  '_person': _person,
  'nametag': Nametag,
  'profile': Profile,
}

export type TemplateName = keyof TemplateMap

export const TEMPLATES = [
  '_person',
  'nametag',
  'profile',
] as const"
`;
